// @sendbird/uikit-react/cjs/chunks/bundle-C0HSA_B2.js
'use strict';

var React = require('react');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefaultCompat(React);

/**
 * user profile goes deep inside the component tree
 * use this context as a short circuit to send in values
 */
var UserProfileContext = React__default.default.createContext({
    disableUserProfile: true,
    isOpenChannel: false,
});
var UserProfileProvider = function (_a) {
    var _b = _a.isOpenChannel, isOpenChannel = _b === void 0 ? false : _b, _c = _a.disableUserProfile, disableUserProfile = _c === void 0 ? false : _c, renderUserProfile = _a.renderUserProfile, onUserProfileMessage = _a.onUserProfileMessage, children = _a.children;
    return (React__default.default.createElement(UserProfileContext.Provider, { value: {
            isOpenChannel: isOpenChannel,
            disableUserProfile: disableUserProfile,
            renderUserProfile: renderUserProfile,
            onUserProfileMessage: onUserProfileMessage,
        } }, children));
};

exports.UserProfileContext = UserProfileContext;
exports.UserProfileProvider = UserProfileProvider;
