// @sendbird/uikit-react/Thread/components/ThreadList.js
import { _ as __assign } from '../../chunks/bundle-BQi9-O76.js';
import React__default from 'react';
import ThreadListItem from './ThreadListItem.js';
import { useThreadContext } from '../context.js';
import { compareMessagesForGrouping } from '../../Channel/utils/compareMessagesForGrouping.js';
import { useSendbirdStateContext } from '../../useSendbirdStateContext.js';
import { isSameDay } from 'date-fns';
import { MessageProvider } from '../../Message/context.js';
import { g as getCaseResolvedReplyType } from '../../chunks/bundle-fHeMHhND.js';
import '../../chunks/bundle-CAdqcqOm.js';
import '../../chunks/bundle-CA85nVxp.js';
import '../../chunks/bundle-bdDHjHH6.js';
import '../../chunks/bundle-cQZ3UZoG.js';
import '../../chunks/bundle-Bwk5KPwK.js';
import '../../ui/DateSeparator.js';
import '../../chunks/bundle-rA4afvKU.js';
import '../../chunks/bundle-DhKQeIfP.js';
import '../../chunks/bundle-D6ZpcPbH.js';
import '../../chunks/bundle-CXdo2kR_.js';
import '../../chunks/bundle-DxTN4_0p.js';
import '@sendbird/chat/groupChannel';
import '../../utils/message/getOutgoingMessageState.js';
import '../../chunks/bundle-CNWAfTVw.js';
import '../../chunks/bundle-BgwzpLeQ.js';
import 'react-dom';
import '../../chunks/bundle-f-SpnOx2.js';
import '../../ui/IconButton.js';
import '../../ui/Button.js';
import '../../ui/Icon.js';
import '@sendbird/uikit-tools';
import '../../withSendbird.js';
import 'css-vars-ponyfill';
import '@sendbird/chat';
import '../../chunks/bundle-etuaqevl.js';
import '@sendbird/chat/openChannel';
import '../../chunks/bundle-DDeWOERD.js';
import '../../chunks/bundle-DCMGp6rH.js';
import '../../chunks/bundle-CwKZj-B6.js';
import '../../ui/SortByRow.js';
import '../../chunks/bundle-Zvm5CMGT.js';
import '../../ui/FileViewer.js';
import '../../chunks/bundle-CFosTNce.js';
import '../../ui/ImageRenderer.js';
import '../../chunks/bundle-DXRRGizV.js';
import '../../chunks/bundle-Gdl-rzQC.js';
import '../../chunks/bundle-BsAIw_77.js';
import '../../chunks/bundle-Due9yzp6.js';
import '../../chunks/bundle-CveCFQYP.js';
import '../../chunks/bundle-0qSTzMXb.js';
import '../../ui/MessageInput.js';
import '../../chunks/bundle-TZQ6dJCG.js';
import 'dompurify';
import '../../chunks/bundle-0nQK2G2O.js';
import '../../chunks/bundle-B6edg1g0.js';
import '../../chunks/bundle-DhGZA5UP.js';
import '../../chunks/bundle-GUbI4JcD.js';
import '../context/types.js';
import '../../chunks/bundle-BrTJ5Zbd.js';
import '../../ui/UserProfile.js';
import '../../sendbirdSelectors.js';
import '../../chunks/bundle-b6TpX_JP.js';
import '../../ui/MessageItemReactionMenu.js';
import '../../ui/ReactionButton.js';
import '../../chunks/bundle-AfQS5DGf.js';
import '../../chunks/bundle-vuBlPTsL.js';
import '../../ui/Loader.js';
import '../../chunks/bundle-BHK1eF5R.js';
import '../../chunks/bundle-F5koPk19.js';
import '../../ui/EmojiReactions.js';
import '../../ui/ReactionBadge.js';
import '../../ui/BottomSheet.js';
import '../../ui/UserListItem.js';
import '../../ui/MutedAvatarOverlay.js';
import '../../ui/Checkbox.js';
import '../../ui/Tooltip.js';
import '../../ui/TooltipWrapper.js';
import '../../chunks/bundle-BNDaXsLn.js';
import '../../ui/TextMessageItemBody.js';
import '../../chunks/bundle-DofK8ySe.js';
import '../../ui/MentionLabel.js';
import '../../ui/LinkLabel.js';
import '../../ui/OGMessageItemBody.js';
import '../../ui/FileMessageItemBody.js';
import '../../ui/TextButton.js';
import '../../ui/ThumbnailMessageItemBody.js';
import '../../chunks/bundle-BqMHiKkl.js';
import '../../ui/UnknownMessageItemBody.js';
import '../../ui/VoiceMessageItemBody.js';
import '../../ui/ProgressBar.js';
import '../../VoicePlayer/useVoicePlayer.js';
import '../../ui/PlaybackTime.js';
import '../../ui/MobileMenu.js';
import '../../chunks/bundle-ZVq6xGXk.js';
import '@sendbird/chat/message';
import '../../chunks/bundle-DkBecO3w.js';
import '../../chunks/bundle-Cf0UgZwi.js';
import '../../chunks/bundle-C3Ty1Yuy.js';
import '../../Message/hooks/useDirtyGetMentions.js';
import '../../chunks/bundle-T6fVCCb7.js';
import '../../chunks/bundle-CMLImr6T.js';
import '../../chunks/bundle-CpjASiz4.js';
import '../../chunks/bundle-CZUYMubi.js';

function ThreadList(_a) {
    var className = _a.className, _b = _a.renderMessage, renderMessage = _b === void 0 ? function (props) { return React__default.createElement(ThreadListItem, __assign({}, props)); } : _b, renderCustomSeparator = _a.renderCustomSeparator, scrollRef = _a.scrollRef, scrollBottom = _a.scrollBottom;
    var config = useSendbirdStateContext().config;
    var userId = config.userId;
    var _c = useThreadContext(), currentChannel = _c.currentChannel, allThreadMessages = _c.allThreadMessages, localThreadMessages = _c.localThreadMessages;
    return (React__default.createElement("div", { className: "sendbird-thread-list ".concat(className) },
        allThreadMessages.map(function (message, idx) {
            var _a;
            var isByMe = ((_a = message === null || message === void 0 ? void 0 : message.sender) === null || _a === void 0 ? void 0 : _a.userId) === userId;
            var prevMessage = allThreadMessages[idx - 1];
            var nextMessage = allThreadMessages[idx + 1];
            // eslint-disable-next-line no-constant-condition
            var _b = compareMessagesForGrouping(prevMessage, message, nextMessage, currentChannel, getCaseResolvedReplyType(config.groupChannel.replyType).upperCase)
                , chainTop = _b[0], chainBottom = _b[1];
            var hasSeparator = !((prevMessage === null || prevMessage === void 0 ? void 0 : prevMessage.createdAt) > 0 && (isSameDay(message === null || message === void 0 ? void 0 : message.createdAt, prevMessage === null || prevMessage === void 0 ? void 0 : prevMessage.createdAt)));
            var handleScroll = function () {
                var current = scrollRef === null || scrollRef === void 0 ? void 0 : scrollRef.current;
                if (current && scrollBottom) {
                    var bottom = current.scrollHeight - current.scrollTop - current.offsetHeight;
                    if (scrollBottom < bottom) {
                        current.scrollTop += bottom - scrollBottom;
                    }
                }
            };
            return (React__default.createElement(MessageProvider, { message: message, isByMe: isByMe, key: message === null || message === void 0 ? void 0 : message.messageId }, renderMessage({
                message: message,
                chainTop: chainTop,
                chainBottom: chainBottom,
                hasSeparator: hasSeparator,
                renderCustomSeparator: renderCustomSeparator,
                handleScroll: handleScroll,
            })));
        }),
        localThreadMessages.map(function (message) {
            var _a;
            var isByMe = ((_a = message === null || message === void 0 ? void 0 : message.sender) === null || _a === void 0 ? void 0 : _a.userId) === userId;
            var handleScroll = function () {
                var current = scrollRef === null || scrollRef === void 0 ? void 0 : scrollRef.current;
                if (current) {
                    var bottom = current.scrollHeight - current.scrollTop - current.offsetHeight;
                    if (scrollBottom < bottom) {
                        current.scrollTop += bottom - scrollBottom;
                    }
                }
            };
            return (React__default.createElement(MessageProvider, { message: message, isByMe: isByMe, key: message === null || message === void 0 ? void 0 : message.messageId }, renderMessage({
                message: message,
                hasSeparator: false,
                renderCustomSeparator: renderCustomSeparator,
                handleScroll: handleScroll,
            })));
        })));
}

export { ThreadList as default };
