// @sendbird/uikit-react/GroupChannelList/components/GroupChannelListUI.js
import { _ as __assign, a as __awaiter, b as __generator } from '../../chunks/bundle-BQi9-O76.js';
import React__default from 'react';
import { useGroupChannelListContext } from '../context.js';
import { G as GroupChannelListUIView } from '../../chunks/bundle-CRcPMrRq.js';
import { GroupChannelPreviewAction } from './GroupChannelPreviewAction.js';
import { useSendbirdStateContext } from '../../useSendbirdStateContext.js';
import { GroupChannelListItem } from './GroupChannelListItem.js';
import { AddGroupChannel } from './AddGroupChannel.js';
import { n as noop } from '../../chunks/bundle-BgwzpLeQ.js';
import '@sendbird/chat/groupChannel';
import '@sendbird/uikit-tools';
import '../../chunks/bundle-BrTJ5Zbd.js';
import '../../chunks/bundle-Soz6sOqr.js';
import '@sendbird/chat';
import '../../chunks/bundle-etuaqevl.js';
import './GroupChannelListHeader.js';
import '../../chunks/bundle-cQZ3UZoG.js';
import '../../chunks/bundle-Bwk5KPwK.js';
import '../../chunks/bundle-bdDHjHH6.js';
import '../../chunks/bundle-CFosTNce.js';
import '../../ui/ImageRenderer.js';
import '../../chunks/bundle-DXRRGizV.js';
import '../../ui/Icon.js';
import '../../chunks/bundle-DhKQeIfP.js';
import '../../ui/Header.js';
import '../../chunks/bundle-f-SpnOx2.js';
import '../../ui/TextButton.js';
import '../../chunks/bundle-rA4afvKU.js';
import '../../ui/IconButton.js';
import '../../EditUserProfile.js';
import '../../EditUserProfile/context.js';
import '../../EditUserProfile/components/EditUserProfileUI.js';
import '../../chunks/bundle-B7BfObV1.js';
import '../../chunks/bundle-CXdo2kR_.js';
import '../../chunks/bundle-DxTN4_0p.js';
import '../../utils/message/getOutgoingMessageState.js';
import '../../chunks/bundle-CNWAfTVw.js';
import '../../chunks/bundle-CA85nVxp.js';
import 'react-dom';
import '../../ui/Button.js';
import '../../withSendbird.js';
import 'css-vars-ponyfill';
import '@sendbird/chat/openChannel';
import '../../chunks/bundle-DDeWOERD.js';
import '../../chunks/bundle-DCMGp6rH.js';
import '../../chunks/bundle-CwKZj-B6.js';
import '../../ui/SortByRow.js';
import '../../ui/Input.js';
import '../../ui/PlaceHolder.js';
import '../../ui/Loader.js';
import '../../chunks/bundle-DvsVgg5f.js';
import '../../chunks/bundle-CMLImr6T.js';
import '@sendbird/chat/message';
import '../../chunks/bundle-DSn10d8F.js';
import '../../chunks/bundle-vuBlPTsL.js';
import '../../chunks/bundle-CAdqcqOm.js';
import '../../chunks/bundle-BHK1eF5R.js';
import '../../chunks/bundle-F5koPk19.js';
import '../../chunks/bundle-C3GXmMjX.js';
import '../../chunks/bundle-AfQS5DGf.js';
import '../../GroupChannel/components/TypingIndicator.js';
import '../../ui/Badge.js';
import '../../ui/ChannelAvatar.js';
import '../../chunks/bundle-GMlM2rlV.js';
import '../../ui/MentionUserLabel.js';
import '../../chunks/bundle-0nQK2G2O.js';
import '../../chunks/bundle-Ba2Z_B16.js';
import '../../CreateChannel.js';
import '../../CreateChannel/components/CreateChannelUI.js';
import '../../chunks/bundle-CBUfFuFR.js';
import '../../sendbirdSelectors.js';
import '../../chunks/bundle-b6TpX_JP.js';
import '../../CreateChannel/components/InviteUsers.js';
import '../../ui/UserListItem.js';
import '../../ui/MutedAvatarOverlay.js';
import '../../ui/Checkbox.js';
import '../../ui/UserProfile.js';
import '../../CreateChannel/components/SelectChannelType.js';

var GroupChannelListUI = function (props) {
    var renderHeader = props.renderHeader, renderChannelPreview = props.renderChannelPreview, renderPlaceHolderError = props.renderPlaceHolderError, renderPlaceHolderLoading = props.renderPlaceHolderLoading, renderPlaceHolderEmptyList = props.renderPlaceHolderEmptyList;
    var _a = useGroupChannelListContext(), onChannelSelect = _a.onChannelSelect, onThemeChange = _a.onThemeChange, allowProfileEdit = _a.allowProfileEdit, typingChannelUrls = _a.typingChannelUrls, groupChannels = _a.groupChannels, initialized = _a.initialized, selectedChannelUrl = _a.selectedChannelUrl, loadMore = _a.loadMore, onUserProfileUpdated = _a.onUserProfileUpdated;
    var _b = useSendbirdStateContext(), stores = _b.stores, config = _b.config;
    var logger = config.logger, isOnline = config.isOnline;
    var sdk = stores.sdkStore.sdk;
    var renderListItem = function (renderProps) {
        var channel = renderProps.item, index = renderProps.index;
        var itemProps = {
            channel: channel,
            tabIndex: index,
            isSelected: channel.url === selectedChannelUrl,
            isTyping: typingChannelUrls.includes(channel.url),
            renderChannelAction: function (props) { return React__default.createElement(GroupChannelPreviewAction, __assign({}, props)); },
            onClick: function () {
                if (isOnline || (sdk === null || sdk === void 0 ? void 0 : sdk.isCacheEnabled)) {
                    logger.info('ChannelList: Clicked on channel:', channel);
                    onChannelSelect(channel);
                }
                else {
                    logger.warning('ChannelList: Inactivated clicking channel item during offline.');
                }
            },
            onLeaveChannel: function () {
                return __awaiter(this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                logger.info('ChannelList: Leaving channel', channel);
                                return [4 /*yield*/, channel.leave()];
                            case 1:
                                _a.sent();
                                logger.info('ChannelList: Leaving channel success');
                                return [2 /*return*/];
                        }
                    });
                });
            },
        };
        if (renderChannelPreview) {
            return (React__default.createElement("div", { key: channel.url, onClick: itemProps.onClick }, renderChannelPreview(itemProps)));
        }
        return React__default.createElement(GroupChannelListItem, __assign({ key: channel.url }, itemProps));
    };
    return (React__default.createElement(GroupChannelListUIView, { renderHeader: renderHeader, renderChannel: renderListItem, renderPlaceHolderError: renderPlaceHolderError, renderPlaceHolderLoading: renderPlaceHolderLoading, renderPlaceHolderEmptyList: renderPlaceHolderEmptyList, onChangeTheme: onThemeChange !== null && onThemeChange !== void 0 ? onThemeChange : noop, allowProfileEdit: allowProfileEdit, onUserProfileUpdated: onUserProfileUpdated !== null && onUserProfileUpdated !== void 0 ? onUserProfileUpdated : noop, channels: groupChannels, onLoadMore: loadMore, initialized: initialized, renderAddChannel: function () { return React__default.createElement(AddGroupChannel, null); } }));
};

export { GroupChannelListUI, GroupChannelListUI as default };
